### PR TITLE
v.0.9.0 release draft

### PR DESCRIPTION
**Pull request by abhiramm7**
In this PR, we support pollutant handling and several other enhancements. The issues we will be patching in this release will be referenced in the PR.

### CODE REVIEW

**Comment by karosc** [comment]
I ran a code coverage report on the tests, which produced the output below.

image

Although all tests are passing, there are a few functions that aren't getting tested including. We probably should add tests for these eventually, but not sure if it's important enough for this release. For what it's worth, my development version of swmm-pandas tests all swmm.toolkit.output functions and is passing with this version of swmm-python

1. output.get_node_series
2. output.get_link_series
3. output.get_system_series
4. output.get_node_attribute
5. output.get_link_attribute
6. solver.swmm_get_mass_balance
7. solver.swmm_build_id

There are also some swiginit functions not called in the following solver classes, but I'm not sure how important they are:
1. solver.RunoffTotals
2. solver.RoutingTotals
3. solver.SubcatchStats
4. solver.PumpStats
5. solver.LinkStats
6. solver.OutfallStats
7. solver.StorageStats
8. solver.NodeStats



Let me know if there anything I can do  to help push this release out! Thanks for all your hard work Abhi 😁!

**Comment by abhiramm7** [comment]
@karosc, this is awesome! Thanks for running the code coverage 😁. I'll create a new issue to port over the tests from swmm-pandas to swmm-python. I was thinking of copying them over directly, but I am going back and forth about adding a new dependency for testing. So, for now, I'll go ahead with the tests we have. 

I think right now. We just need a review. @karosc, if you have some time, can you go over the PR?

**Comment by abhiramm7** [comment]
I went over the checklist, I think we are good to go. Nothing is jumping out to me. @michaeltryby and @jennwuu if you have a moment can you go over the PR?

**Comment by jennwuu** [comment]
Hey @abhiramm7 thank your patience, can you also bump numpy version in the test requirement file as well?

File: swmm-toolkit/setup.py [hunk 1]
```diff
@@ -116,10 +116,10 @@ def exclude_files(cmake_manifest):
         "Operating System :: POSIX :: Linux",
         "Operating System :: MacOS",
         "License :: CC0 1.0 Universal (CC0 1.0) Public Domain Dedication",
-        "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
         "Programming Language :: C",
         "Development Status :: 4 - Beta",
```
Comment (jennwuu) [inline swmm-toolkit/setup.py:124]: Is development still in BETA? Can we remove this?

File: swmm-toolkit/src/swmm/toolkit/__init__.py [hunk 2]
```diff
@@ -29,13 +29,16 @@
...
 
 
 # Adds directory containing swmm libraries to path
 if platform.system() == "Windows":
     libdir = os.path.join(os.path.dirname(__file__), "../../swmm_toolkit")
 
     if hasattr(os, 'add_dll_directory'):
-        os.add_dll_directory(libdir)
+       if 'conda' in sys.version:
+           os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE']="1"
```
Comment (jennwuu) [inline swmm-toolkit/src/swmm/toolkit/__init__.py:41]: Just minor format comment, can you add space between the = and the value?

Comment (jennwuu) [reply swmm-toolkit/src/swmm/toolkit/__init__.py:41]: is conda case sensitive in sys.version? would it make sense to make it all lowercase before checking?

**Review by jennwuu — CHANGES_REQUESTED** [review]
Minor comment about the format. The tests make sense to me.

Do we have a test for swmm_build_id?

**Comment by jennwuu** [comment]
> I ran a code coverage report on the tests, which produced the output below.
> 
> image
> 
> Although all tests are passing, there are a few functions that aren't getting tested (listed below). We probably should add tests for these, but is it important enough for this release? For what it's worth, my development version of swmm-pandas tests all swmm.toolkit.output functions and is passing with this version of swmm-python
> 
> 1. output.get_node_series
> 2. output.get_link_series
> 3. output.get_system_series
> 4. output.get_node_attribute
> 5. output.get_link_attribute
> 6. solver.swmm_get_mass_balance
> 7. solver.swmm_build_id
> 
> There are also some swiginit functions not called in the following solver classes, but I'm not sure how important they are:
> 
> 1. solver.RunoffTotals
> 2. solver.RoutingTotals
> 3. solver.SubcatchStats
> 4. solver.PumpStats
> 5. solver.LinkStats
> 6. solver.OutfallStats
> 7. solver.StorageStats
> 8. solver.NodeStats
> 
> Let me know if there anything I can do to help push this release out! Thanks for all your hard work Abhi 😁!

Wow, just noticed this. Thank you for checking. Feel free to add any additional unit tests to complete the code coverage. This PR looks good to me overall. I can give approval after minor adjustment as per PR feedback.

Thanks!

File: swmm-toolkit/setup.py (hunk 1 above)
Comment (abhiramm7) [reply swmm-toolkit/setup.py:124]: Yeah, let's do it. I remove it and update it to production/stable.

File: swmm-toolkit/src/swmm/toolkit/__init__.py (hunk 2 above)
Comment (abhiramm7) [reply swmm-toolkit/src/swmm/toolkit/__init__.py:41]: Good call @jennwuu, I discovered on Stackoverflow that this `conda` check is only valid for python < 3.6 [1]. We no longer support python 3.6. I am updating this `conda` check to be in line with python 3.7 and greater. This patch is adopted from https://stackoverflow.com/questions/47608532/how-to-detect-from-within-python-whether-packages-are-managed-with-conda/47610844#47610844

1 : https://stackoverflow.com/questions/21282363/any-way-to-tell-if-users-python-environment-is-anaconda

Comment (abhiramm7) [reply swmm-toolkit/src/swmm/toolkit/__init__.py]: I added a `conda` test to make sure we are able to install swmm-toolkit via `conda `python. The above addition seems to be working.

Comment (abhiramm7) [reply swmm-toolkit/src/swmm/toolkit/__init__.py]: https://github.com/OpenWaterAnalytics/swmm-python/runs/7355904560?check_suite_focus=true - Test for conda

**Comment by abhiramm7** [comment]
@jennwuu I updated the setup.py and added a test for checking conda install. I think we are good to go. Let me know what you think? :)

**Comment by karosc** [comment]
Nice find on conda-meta @abhiramm7. I confirmed that this method works for both base and custom conda environments even though[ this stack overflow comment says otherwise](https://stackoverflow.com/questions/47608532/how-to-detect-from-within-python-whether-packages-are-managed-with-conda/47610844#comment104657962_47610844).

**Pull request merged**